tdp/core/collection/collection.py

```diff
@@ -125,9 +117,9 @@ def schema_directory(self) -> Path:
         return self._path / SCHEMA_VARS_DIRECTORY_NAME
 
     @property
-    def dag_nodes(self) -> list[TDPLibDagNodeModel]:
-        """List of DAG files in the YAML format."""
-        return self._dag_nodes  # TODO: should return a generator
+    def dag_nodes(self) -> Generator[TDPLibDagNodeModel, None, None]:
+        """Read DAG nodes."""
+        return parse_dag_directory(self.dag_directory)
```

Comment:
You change the logic here, with the new logic, every call to `dag_nodes` read the filesystem which is not what we want. The filesystem should be read once when we populate `Collections` and keep all information read in memory.

Reply:
I did this on purpose. `Collections` is indeed calling `Collection.dag_nodes` only once to initialize itself.

IMO, the `Collection` class should be used to get the **current** collection content. I would like to accomplish the same for the other properties. I planed to do that in another PR but it may makes more sense to do everything at once.

Reply:
That way, we would be able to use the `Collection` class to update the `Collections` object when needed without having to recreate each `Collection` objects.

Reply:
haha now you want to have a `Collection` which have content updated for each function call whereas you are the one who wanted to have immutable classes so I don't understand.

Reply:
Yeah, I kinda changed my mind for `Collection` 💁🏻‍♂️

Reply:
Fixed it in the new version : I removed the `@property` (except for directories, name and path) and replace them with methods `read_dag_nodes`, `read_playbooks` and `read_schemas`.